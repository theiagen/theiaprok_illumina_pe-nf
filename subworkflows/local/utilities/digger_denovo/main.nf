#!/usr/bin/env nextflow

/*
========================================================================================
    DIGGER DE NOVO ASSEMBLY SUBWORKFLOW
========================================================================================
    Testing conversion from our WDL
----------------------------------------------------------------------------------------
*/

include { SPADES         } from '../../../../modules/local/assembly/spades/main'
include { MEGAHIT        } from '../../../../modules/local/assembly/megahit/main'
include { SKESA          } from '../../../../modules/local/assembly/skesa/main'
include { BWA_MEM        } from '../../../../modules/local/alignment/bwa/main'
include { PILON          } from '../../../../modules/local/polishing/pilon/main'
include { FILTER_CONTIGS } from '../../../../modules/local/quality_control/contig_filtering/shovilter/main'

workflow DIGGER_DENOVO {
    
    take:
    ch_reads        // channel: [ val(meta), [ reads ] ]
    
    main:
    
    ch_versions = Channel.empty()
    
    ch_assembly = Channel.empty()
    ch_assembly_gfa = Channel.empty()
    
    if (params.assembler == 'spades') {
        SPADES (
            ch_reads,
            params.kmers ?: [],
            params.spades_type ?: 'isolate'
        )
        ch_assembly = SPADES.out.assembly_fasta
        ch_assembly_gfa = SPADES.out.assembly_gfa
        ch_versions = ch_versions.mix(SPADES.out.versions)
    }
    else if (params.assembler == 'megahit') {
        MEGAHIT (
            ch_reads,
            params.kmers ?: ''
        )
        ch_assembly = MEGAHIT.out.assembly_fasta
        ch_versions = ch_versions.mix(MEGAHIT.out.versions)
    }
    else if (params.assembler == 'skesa') {
        SKESA (
            ch_reads
        )
        ch_assembly = SKESA.out.assembly_fasta
        ch_versions = ch_versions.mix(SKESA.out.versions)
    }
    else {
        error "Invalid assembler specified: ${params.assembler}. Valid options: spades, megahit, skesa"
    }
    
    ch_polished_assembly = ch_assembly
    ch_pilon_changes = Channel.empty()
    ch_pilon_vcf = Channel.empty()
    
    if (params.call_pilon) {
        // Prepare reads and assembly for BWA
        ch_bwa_input = ch_reads
            .join(ch_assembly)
            .map { meta, reads, assembly -> 
                [ meta, reads, assembly ]
            }
        
        BWA_MEM (
            ch_bwa_input
        )
        ch_versions = ch_versions.mix(BWA_MEM.out.versions)
        
        // Prepare input for Pilon
        ch_pilon_input = ch_assembly
            .join(BWA_MEM.out.bam)
            .join(BWA_MEM.out.bai)
            .map { meta, assembly, bam, bai ->
                [ meta, assembly, bam, bai ]
            }
        
        PILON (
            ch_pilon_input
        )
        ch_polished_assembly = PILON.out.improved_assembly
        ch_pilon_changes = PILON.out.changes
        ch_pilon_vcf = PILON.out.vcf
        ch_versions = ch_versions.mix(PILON.out.versions)
    }
    
    ch_final_assembly = ch_polished_assembly
    ch_filter_metrics = Channel.empty()
    
    if (params.run_filter_contigs) {
        FILTER_CONTIGS (
            ch_polished_assembly
        )
        ch_final_assembly = FILTER_CONTIGS.out.filtered_contigs
        ch_filter_metrics = FILTER_CONTIGS.out.filter_metrics
        ch_versions = ch_versions.mix(FILTER_CONTIGS.out.versions)
    }
    
    emit:
    assembly           = ch_final_assembly         // channel: [ val(meta), path(assembly) ]
    contigs_gfa        = ch_assembly_gfa          // channel: [ val(meta), path(gfa) ]
    filter_metrics     = ch_filter_metrics        // channel: [ val(meta), path(metrics) ]
    pilon_changes      = ch_pilon_changes         // channel: [ val(meta), path(changes) ]
    pilon_vcf          = ch_pilon_vcf             // channel: [ val(meta), path(vcf) ]
    versions           = ch_versions              // channel: [ versions.yml ]
}