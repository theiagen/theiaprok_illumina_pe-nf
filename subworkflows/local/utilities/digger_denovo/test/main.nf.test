nextflow_workflow {

    name "Test Subworkflow DIGGER_DENOVO"
    script "../main.nf"
    workflow "DIGGER_DENOVO"

    // Spades doesn't assemble the sarscov2 test data, so we use a different test set of prokaryotes
    // for the spades assembler test.
    // The sarscov2 test data is used for megahit and skesa assembler tests

    test("digger_denovo - spades assembler prokaryotes") {

        when {
            params {
                assembler = 'spades'
                kmers = '21, 33, 55'
                spades_type = 'isolate'
                call_pilon = false
                run_filter_contigs = false
            }
            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test', single_end:false ], // meta map
                    [
                        file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_1.fastq.gz", checkIfExists: true),
                        file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_2.fastq.gz", checkIfExists: true)
                    ]
                ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    file(workflow.out.assembly[0][1]).name,
                    file(workflow.out.contigs_gfa[0][1]).name,
                    workflow.out.filter_metrics,
                    workflow.out.pilon_changes,
                    workflow.out.pilon_vcf,
                    workflow.out.versions).match() }
            )
        }
    }

    test("digger_denovo - megahit assembler sarscov2") {

        when {
            params {
                assembler = 'megahit'
                kmers = '21,33,55'
                call_pilon = false
                run_filter_contigs = false
            }
            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test', single_end:false ], // meta map
                    [
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_1.fastq.gz', checkIfExists: true),
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_2.fastq.gz', checkIfExists: true)
                    ]
                ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    file(workflow.out.assembly[0][1]).name,
                    workflow.out.contigs_gfa,
                    workflow.out.filter_metrics,
                    workflow.out.pilon_changes,
                    workflow.out.pilon_vcf,
                    workflow.out.versions).match() }
            )
        }
    }

    test("digger_denovo - skesa assembler sarscov2") {

        when {
            params {
                assembler = 'skesa'
                call_pilon = false
                run_filter_contigs = false
            }
            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test', single_end:false ], // meta map
                    [
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_1.fastq.gz', checkIfExists: true),
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_2.fastq.gz', checkIfExists: true)
                    ]
                ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    file(workflow.out.assembly[0][1]).name,
                    workflow.out.contigs_gfa,
                    workflow.out.filter_metrics,
                    workflow.out.pilon_changes,
                    workflow.out.pilon_vcf,
                    workflow.out.versions).match() }
            )
        }
    }

    test("digger_denovo - full pipeline with pilon and filtering") {

        when {
            params {
                assembler = 'spades'
                kmers = '21, 33, 55'
                spades_type = 'isolate'
                call_pilon = true
                run_filter_contigs = true
            }
            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test', single_end:false ], // meta map
                    [
                        file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_1.fastq.gz", checkIfExists: true),
                        file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_2.fastq.gz", checkIfExists: true)
                    ]
                ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    file(workflow.out.assembly[0][1]).name,
                    file(workflow.out.contigs_gfa[0][1]).name,
                    file(workflow.out.filter_metrics[0][1]).name,
                    file(workflow.out.pilon_changes[0][1]).name,
                    file(workflow.out.pilon_vcf[0][1]).name,
                    workflow.out.versions).match() }
            )
        }
    }

}