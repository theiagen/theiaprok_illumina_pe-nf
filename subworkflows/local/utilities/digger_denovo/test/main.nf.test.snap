{
    "digger_denovo - megahit assembler": {
        "content": [
            "test_megahit_contigs.fa",
            [
                
            ],
            [
                
            ],
            [
                
            ],
            [
                
            ],
            [
                "versions.yml:md5,44a194933db72abb1740ea50740f176e"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.4"
        },
        "timestamp": "2025-06-23T14:36:16.386416531"
    },
    "digger_denovo - skesa assembler": {
        "content": [
            "test_skesa_contigs.fa",
            [
                
            ],
            [
                
            ],
            [
                
            ],
            [
                
            ],
            [
                "versions.yml:md5,f7c83145e4e4df79b00294e5f3cc9bf8"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.4"
        },
        "timestamp": "2025-06-23T14:40:26.447883072"
    },
    "digger_denovo - spades assembler": {
        "content": [
            "test_isolatespades_contigs.fa",
            "test_isolatespades_contigs.gfa",
            [
                
            ],
            [
                
            ],
            [
                
            ],
            [
                "versions.yml:md5,6b7ae1c44a246671b59aa7d7831f667c"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.4"
        },
        "timestamp": "2025-06-23T14:40:00.454566148"
    },
    "digger_denovo - full pipeline with pilon and filtering": {
        "content": [
            "test_filtered_contigs.fasta",
            "test_isolatespades_contigs.gfa",
            "test_filtering_metrics.txt",
            "test_pilon.changes",
            "test_pilon.vcf",
            [
                "versions.yml:md5,2ddcb7b2e649d5a42bdde8870e7bb82d",
                "versions.yml:md5,6b7ae1c44a246671b59aa7d7831f667c",
                "versions.yml:md5,990c85ed7fc1c013e309e6e18451ff3e",
                "versions.yml:md5,c1c66d929d0d43d9db27d2978db4fbac"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.4"
        },
        "timestamp": "2025-06-23T14:41:43.422189844"
    }
}